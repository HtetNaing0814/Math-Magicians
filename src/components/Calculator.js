/* eslint-disable no-eval */

import React, { useState } from 'react';
import Display from './Display';
import Button from './Button';

const Calculator = () => {
  const [value, setValue] = useState('0');
  const ops = ['+', '-', '*', '/', '%'];

  const updateValue = (newValue) => {
    setValue(value === '0' ? newValue : value + newValue);
  };

  const equalTo = () => {
    setValue(ops.includes(value[value.length - 1]) ? value : `${eval(value)}`);
  };

  const clearAll = () => {
    setValue('0');
  };

  const deleteLastDigit = () => {
    setValue(value.length === 1 ? '0' : value.slice(0, -1));
  };

  const operations = (newValue) => {
    setValue(value === '0' || ops.includes(value[value.length - 1]) ? value : value + newValue);
  };

  const decimal = (newValue) => {
    let index = '';
    for (let i = 0; i < value.length; i += 1) {
      if (ops.includes(value[i])) {
        index = value.lastIndexOf(value[i]) + 1;
      }
    }
    const temp = value.substring(index);
    setValue(temp.includes('.') ? value : value + newValue);
  };

  return (
    <div id="calculator-container">
      <p>Let&apos;s perform some MATHGIC !!!</p>
      <div id="calculator">
        <Display value={value} />
        <br />
        <Button label="AC" updateValue={() => clearAll()} />
        <Button label="DEL" updateValue={() => deleteLastDigit()} />
        <Button label="%" updateValue={() => operations('%')} />
        <Button label="/" bgColor="orange" updateValue={() => operations('/')} />
        <br />
        <Button label="7" updateValue={() => updateValue('7')} />
        <Button label="8" updateValue={() => updateValue('8')} />
        <Button label="9" updateValue={() => updateValue('9')} />
        <Button label="*" bgColor="orange" updateValue={() => operations('*')} />
        <br />
        <Button label="4" updateValue={() => updateValue('4')} />
        <Button label="5" updateValue={() => updateValue('5')} />
        <Button label="6" updateValue={() => updateValue('6')} />
        <Button label="-" bgColor="orange" updateValue={() => operations('-')} />
        <br />
        <Button label="1" updateValue={() => updateValue('1')} />
        <Button label="2" updateValue={() => updateValue('2')} />
        <Button label="3" updateValue={() => updateValue('3')} />
        <Button label="+" bgColor="orange" updateValue={() => operations('+')} />
        <br />
        <Button label="0" updateValue={() => updateValue('0')} width="200px" />
        <Button label="." updateValue={() => decimal('.')} />
        <Button label="=" bgColor="orange" updateValue={() => equalTo()} />
      </div>
    </div>
  );
};

export default Calculator;
