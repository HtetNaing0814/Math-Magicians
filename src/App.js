import React from 'react';
import {
  BrowserRouter as Router,
  Routes,
  Route,
} from 'react-router-dom';
import Calculator from './components/Calculator';
import Nav from './components/Nav';
import Home from './components/Home';
import Quote from './components/Quote';

const App = () => (
  <>
    <Router>
      <Nav />
      <Routes>
        <Route path="/quote" element={<Quote />} />
        <Route path="/calculator" element={<Calculator />} />
        <Route path="/" element={<Home />} />
      </Routes>
    </Router>
  </>
);

export default App;
