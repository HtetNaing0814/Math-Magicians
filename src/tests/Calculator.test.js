import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import Calculator from '../components/Calculator';
import App from '../App';

const labels = ['AC', 'DEL', '%', '/', '*', '-', '+', '=', '.', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
const numbers = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
const ops = ['AC', 'DEL', '%', '/', '*', '-', '+', '=', '.'];

describe('Render the components of Calculator', () => {
  test('renders Display component', () => {
    render(<Calculator />);
    expect(screen.getByRole('textbox')).toMatchSnapshot(); // Display component
  });

  labels.forEach((label) => {
    test(`renders ${label} Button component`, () => {
      render(<Calculator />);
      expect(screen.getByText(label)).toMatchSnapshot(); // All of the button components
    });
  });
});

describe('Render the rest of the components', () => {
  test('Render routes', async () => {
    render(<App />);

    expect(screen.getByText(/Math Magicians/i)).toMatchSnapshot(); // Nav component

    expect(screen.getByText(/Welcome to the World of Mathgic!/i)).toMatchSnapshot(); // Home component

    fireEvent.click(screen.getByText(/Calculator/i));
    expect(screen.getByText(/Let's perform some MATHGIC !!!/i)).toMatchSnapshot(); // Calculator component

    fireEvent.click(screen.getByText(/Quote/i));
    expect(screen.getByText(/Mathematics is not about numbers, equations, computations or algorithms: it is about understanding. - William Paul Thurston/i)).toMatchSnapshot(); // Quote component
  });
});

describe('Operations', () => {
  numbers.forEach((number) => {
    test(`clicks ${number} Button component`, () => {
      render(<Calculator />);
      fireEvent.click(screen.getByText(number));
      expect(screen.getByRole('textbox').value).toBe(number);
    });
  });

  ops.forEach((op) => {
    test(`clicks ${op} Button component`, () => {
      render(<Calculator />);
      const initialValue = screen.getByRole('textbox').value;
      fireEvent.click(screen.getByText(op));
      const { value } = screen.getByRole('textbox');
      if (initialValue !== '0') {
        expect(value).toBe(op);
      }
    });
  });

  test('2.5+5.5-3*3/2%5', () => {
    render(<Calculator />);
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('.'));
    fireEvent.click(screen.getByText('5'));
    fireEvent.click(screen.getByText('+'));
    fireEvent.click(screen.getByText('5'));
    fireEvent.click(screen.getByText('.'));
    fireEvent.click(screen.getByText('5'));
    fireEvent.click(screen.getByText('-'));
    fireEvent.click(screen.getByText('3'));
    fireEvent.click(screen.getByText('*'));
    fireEvent.click(screen.getByText('3'));
    fireEvent.click(screen.getByText('/'));
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('%'));
    fireEvent.click(screen.getByText('5'));
    fireEvent.click(screen.getByText('='));
    expect(screen.getByRole('textbox').value).toBe('3.5');
  });

  test('operators duplicate validation', () => {
    render(<Calculator />);
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('+'));
    fireEvent.click(screen.getByText('+'));
    fireEvent.click(screen.getByText('-'));
    fireEvent.click(screen.getByText('*'));
    fireEvent.click(screen.getByText('/'));
    expect(screen.getByRole('textbox').value).toBe('2+');
  });

  test('decimal duplicate validation', () => {
    render(<Calculator />);
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('.'));
    fireEvent.click(screen.getByText('5'));
    fireEvent.click(screen.getByText('.'));
    fireEvent.click(screen.getByText('*'));
    fireEvent.click(screen.getByText('/'));
    expect(screen.getByRole('textbox').value).toBe('2.5*');
    fireEvent.click(screen.getByText('4'));
    fireEvent.click(screen.getByText('.'));
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('.'));
    fireEvent.click(screen.getByText('='));
    expect(screen.getByRole('textbox').value).toBe('10.5');
  });

  test('AC and DEL', () => {
    render(<Calculator />);
    fireEvent.click(screen.getByText('9'));
    fireEvent.click(screen.getByText('0'));
    fireEvent.click(screen.getByText('1'));
    fireEvent.click(screen.getByText('DEL'));
    expect(screen.getByRole('textbox').value).toBe('90');
    fireEvent.click(screen.getByText('AC'));
    expect(screen.getByRole('textbox').value).toBe('0');
  });
});
