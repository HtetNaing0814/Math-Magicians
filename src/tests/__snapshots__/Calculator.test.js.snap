// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render the components of Calculator renders % Button component 1`] = `
<button
  label="%"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  %
</button>
`;

exports[`Render the components of Calculator renders * Button component 1`] = `
<button
  label="*"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: orange; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  *
</button>
`;

exports[`Render the components of Calculator renders + Button component 1`] = `
<button
  label="+"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: orange; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  +
</button>
`;

exports[`Render the components of Calculator renders . Button component 1`] = `
<button
  label="."
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  .
</button>
`;

exports[`Render the components of Calculator renders / Button component 1`] = `
<button
  label="/"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: orange; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  /
</button>
`;

exports[`Render the components of Calculator renders = Button component 1`] = `
<button
  label="="
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: orange; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  =
</button>
`;

exports[`Render the components of Calculator renders - Button component 1`] = `
<button
  label="-"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: orange; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  -
</button>
`;

exports[`Render the components of Calculator renders 0 Button component 1`] = `
<button
  label="0"
  style="width: 200px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  0
</button>
`;

exports[`Render the components of Calculator renders 1 Button component 1`] = `
<button
  label="1"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  1
</button>
`;

exports[`Render the components of Calculator renders 2 Button component 1`] = `
<button
  label="2"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  2
</button>
`;

exports[`Render the components of Calculator renders 3 Button component 1`] = `
<button
  label="3"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  3
</button>
`;

exports[`Render the components of Calculator renders 4 Button component 1`] = `
<button
  label="4"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  4
</button>
`;

exports[`Render the components of Calculator renders 5 Button component 1`] = `
<button
  label="5"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  5
</button>
`;

exports[`Render the components of Calculator renders 6 Button component 1`] = `
<button
  label="6"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  6
</button>
`;

exports[`Render the components of Calculator renders 7 Button component 1`] = `
<button
  label="7"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  7
</button>
`;

exports[`Render the components of Calculator renders 8 Button component 1`] = `
<button
  label="8"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  8
</button>
`;

exports[`Render the components of Calculator renders 9 Button component 1`] = `
<button
  label="9"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  9
</button>
`;

exports[`Render the components of Calculator renders AC Button component 1`] = `
<button
  label="AC"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  AC
</button>
`;

exports[`Render the components of Calculator renders DEL Button component 1`] = `
<button
  label="DEL"
  style="width: 100px; height: 50px; padding: 5px; font-size: 24px; background-color: white; cursor: pointer; border: 1px solid gray;"
  type="button"
>
  DEL
</button>
`;

exports[`Render the components of Calculator renders Display component 1`] = `
<input
  id="display"
  readonly=""
  type="text"
  value="0"
/>
`;

exports[`Render the rest of the components Render routes 1`] = `
<h1>
  Math Magicians ðŸ”®
</h1>
`;

exports[`Render the rest of the components Render routes 2`] = `
<h1>
  Welcome to the World of Mathgic! âœ¨
</h1>
`;

exports[`Render the rest of the components Render routes 3`] = `
<p>
  Let's perform some MATHGIC !!!
</p>
`;

exports[`Render the rest of the components Render routes 4`] = `
<div
  id="quote"
>
  Mathematics is not about numbers, equations, computations
  <br />
  <br />
   
  or algorithms: it is about understanding. - William Paul Thurston
</div>
`;
